### ISSUE TITLE
Russian filenames are corrupt in Windows GUI

### ISSUE THREAD

**Issue by vovkkk** [bleachbit/bleachbit#14]
Python 2 raises localized exceptions using system encoding, e.g. if you use Russian Windows you may spot in log such error:

```
error: (5, 'MoveFileEx', '\xce\xf2\xea\xe0\xe7\xe0\xed\xee \xe2 \xe4\xee\xf1\xf2\xf3\xef\xe5.')
```

The mystery part was encoded in cp1251, and it’s supposed to be

```
error: (5, 'MoveFileEx', 'Отказано в доступе.')
```

[`locale.getpreferredencoding()`](https://docs.python.org/2.7/library/locale.html#locale.getpreferredencoding) tries to _guess_ the system encoding, and from my experience it guesses correctly on Windows, here is [example](https://github.com/aziz/SublimeFileBrowser/blob/master/dired.py#L534-L555) of my implementation for that stuff in another project (note ST3 in example’s code means Py3), I do not imply that how it must be done, it’s just an example which works, perhaps it might be done better.

**Comment by az0** [comment]
To test this, how do you force Python to show localized exceptions? On my Windows system I tried environment variables and settings in the Windows Control Panel, but the exceptions are in English.

**Comment by vovkkk** [comment]
I do not know, it seems very inconsistent: sometimes exseption contains actual message, sometimes message is an empty string.

For example, I can get following weirdness in IDLE 2.7.8:

![2015-09-13 08-43-57 python 2 7 8 shell](https://cloud.githubusercontent.com/assets/1975298/9834955/f951ee50-59f3-11e5-9383-14b31a422c45.png)

In first case it properly say that there is sintax error in filename (since backslashes are not escaped)
in second case it say nothing, while should complain that directory is not empty.

Also, I’m not really familiar with codebase of BleachBit, but I have a feeling that `MoveFileEx` has some relation to pywin or ctypes, no?

**Comment by vovkkk** [comment]
# How to reproduce for pywin:

1. Make sure that you use some language that contains non-ascii characters (Greek, Russian, etc.) as a **GUI language** (set it in Control panel) and choose that language for application that do not support Unicode:

    ![2015-09-13 11-08-59](https://cloud.githubusercontent.com/assets/1975298/9835337/53f5322c-5a08-11e5-87df-7ae24590f0ae.png)
    ![2015-09-13 11-09-37](https://cloud.githubusercontent.com/assets/1975298/9835338/59548150-5a08-11e5-80c2-7162316c8b6e.png)

2. Create some file and make it unaccessible for current user (Properties → Security).
3. Try to use `MoveFileEx` on that file.

    ```python
    import win32api, win32con

    win32api.MoveFileEx('C:\\Users\\vova\\Desktop\\1', None, win32con.MOVEFILE_DELAY_UNTIL_REBOOT)
    ```

# Results

* Russian: `error: (5, 'MoveFileEx', '\xce\xf2\xea\xe0\xe7\xe0\xed\xee \xe2 \xe4\xee\xf1\xf2\xf3\xef\xe5.')`

    ![2015-09-13 10-04-31 python 2 7 8 shell](https://cloud.githubusercontent.com/assets/1975298/9835344/6a2e9402-5a08-11e5-8ede-d895e08ed35c.png)


* Greek: `error: (5, 'MoveFileEx', '\xc4\xe5\xed \xe5\xf0\xe9\xf4\xf1\xdd\xf0\xe5\xf4\xe1\xe9 \xe7 \xf0\xf1\xfc\xf3\xe2\xe1\xf3\xe7.')`

    ![2015-09-13 11-06-11 python 2 7 8 shell](https://cloud.githubusercontent.com/assets/1975298/9835346/6e1b6e96-5a08-11e5-8bfe-454afbb09740.png)

**Comment by az0** [comment]
Is this still an issue with BleachBit 2.0 or greater? There were many fixes for encoding in BleachBit 2.0

**Comment by vovkkk** [comment]
Apparently it is not an issue in 2.0, but there is a similar issue with paths:
![2018-04-02_14-04-23](https://user-images.githubusercontent.com/1975298/38186769-caad7f6e-367e-11e8-9bc4-50c65a0b5c5d.png)

![2018-04-02_14-41-08](https://user-images.githubusercontent.com/1975298/38187944-f6440d6e-3683-11e8-901f-f033fc50e815.png)


`Новая папка (2)\Новый текстовый документ.txt`

**Comment by az0** [comment]
Thank you for the screenshot. I change the title of this ticket to: **Russian filenames are corrupt in Windows GUI**

I will try your instructions to change the Windows settings to reproduce this issue

**Comment by az0** [comment]
I've had trouble reproducing this. I might just need mor eitme. I'll re-target to another milestone to allow more time.

**Comment by az0** [comment]
@vovkkk  I still have not been able to reproduce this, but there have been many related fixes and changes. Would you please test [build 1237](https://ci.appveyor.com/project/az0/bleachbit/builds/26844010/artifacts)? This build has some special encoding fixes developed by a contractor focusing on encoding issues. These fixes are not yet merged into the master branch, so they are _not_ in most of the [CI builds](https://ci.bleachbit.org) yet. In other words, please don't test a newer build at this time: please test build 1237.

**Comment by vovkkk** [comment]
There are still issue in recognizing custom paths, even with Latin letters:

![2019-08-24_11-09-40](https://user-images.githubusercontent.com/1975298/63632468-ad68d700-c660-11e9-8b29-6c893b293bd4.png)
![2019-08-24_11-18-51](https://user-images.githubusercontent.com/1975298/63632485-00428e80-c661-11e9-8630-1b7d3183d444.png)

```
Новая папка (2)
pasgan ùr
```



Messages are improved


![2019-08-24_10-53-07](https://user-images.githubusercontent.com/1975298/63632508-63342580-c661-11e9-8c38-10db493dcb4c.png)

**Comment by az0** [comment]
Recently there was much work to update from Python 2.7 to Python 3 on Windows. While this wasn't directly targeted at this bug, it did have a major impact on how the system handles Unicode, so it might have helped.

Would you please test [build 3.9.0.1540](https://ci.bleachbit.org/?prefix=dl/3.9.0.1540/) or newer?

**Comment by az0** [comment]
I am closing because (1) no response after some time and (2) it is likely there is an improvement.

If you have the same bug with BleachBit 3.9 or higher, re-open this bug. If the symptoms are different, open a new bug ticket.

**Comment by az0** [comment]
The digitally-signed Windows installer is now available from the page [BleachBit 3.9.0 beta](https://www.bleachbit.org/news/bleachbit-390-beta)
